src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,88 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
+
+            assert(retStructDesc.passedInRegisters);
+
+            // In case the return type is a two-register-return, the native return type should be a struct
+            assert(varTypeIsStruct(compiler->info.compRetNativeType));
+
+            assert(retStructDesc.eightByteCount == CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+            unsigned __int8 offset0 = 0;
+            unsigned __int8 offset1 = 0;
+
+            var_types type0 = TYP_UNKNOWN;
+            var_types type1 = TYP_UNKNOWN;
+
+            // Set the GC-ness of the struct return registers.
+            getStructTypeOffset(retStructDesc, &type0, &type1, &offset0, &offset1);
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, type0);
+            gcInfo.gcMarkRegPtrVal(REG_INTRET_1, type1);
+        }
+        // ... handling return in implicit RetBuf and returning the RetBuf address in REG_INTRET...
+        else
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (compiler->compMethodReturnsRetBufAddr())
+        {
+            // This is for returning in an implicit RetBuf.
+            // If the address of the buffer is returned in REG_INTRET, mark the content of INTRET as ByRef.
+
+            // In case the return is in an implicit RetBuf, the native return type should be a struct
+            assert(varTypeIsStruct(compiler->info.compRetNativeType));
+
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, TYP_BYREF);
+        }
+        // ... all other cases.
+        else
+        {
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+            // For System V the above if statement (if (compiler->compMethodReturnsRetBufAddr()))
+            // should always evaluate to true since the ABI requires to return the 
+            // address of the RetBuf in REG_INTRET.
+            assert(compiler->info.compRetBuffArg == BAD_VAR_NUM);
+
+            // The compRetNativeType should not be a struct type if execution is here on System V - the struct
+            // type return cases are handled in the two if statements above.
+            assert(!varTypeIsStruct(compiler->info.compRetNativeType));
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+
+            // For Windows we can't make such assertions since we generate code for returning of 
+            // the RetBuf in REG_INTRET only when the ProfilerHook is enabled. Otherwise
+            // compRetNativeType could be TYP_STRUCT.
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, compiler->info.compRetNativeType);
+        }
+    }
 
     regNumber regGSCheck;
     if (!pushReg)
     {
         // Non-tail call: we can use any callee trash register that is not
-        // a return register or contain 'this' pointer (keep alive this), since 
+        // a return register (note on Amd64 System V RDX is an arg register -
+        // REG_ARG_2 - as well as return register for two-register-returned structs) 
+        // or contain 'this' pointer (keep alive this), since 
```

Comment:
The parenthesized phrase breaks up this sentence so that it is difficult to parse. I would add the note after the complete sentence.